net: dhcpv4: set sntp server via dhcpv4 option

maass-hamburg opened the pull request:
This adds the option to use dhcpv4 option 42 (NTP Server) to set the ip address of sntp. It also provides changes in init_clock_sntp of net_config, so the ntp server, that is announced via dhcpv4 options,  can be used at startup.

jukkar commented on subsys/net/lib/sntp/Kconfig:
```diff
@@ -19,6 +19,11 @@ config SNTP_UNCERTAINTY
 	  Enabling this option uses more resources (memory) and is not normally
 	  needed.
 
+config SNTP_GLOBAL_ADDR
+	bool "Use global IP address"
+	help
+	  Be able to set a global server IP address for the SNTP client.
```
I am not really convinced that we want to do this. The SNTP library uses the address that is given to it so it does not need to store the address at all. This change just introduces complexity to the system as we need to administrate the server address.
Beside, if we have multiple network interfaces that uses DHCP, then each one could have their own SNTP server. So we could store the received SNTP server address into `net_if`, and then application could get it from there. Then we do not need to have these extra APIs at all.

maass-hamburg committed:
net: ip: add function to set default port

This adds a function, to set the default port
of a sockaddr, if the port had not been set.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
net: lib: sntp_simple: add a function to use with ip address

This adds a function to use sntp_simple with a
ip address directly.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
net: config: use dhcpv4 option for sntp

Allow the use of the NTP server address, set by
dhcpv4 option, by the net_init_clock_via_sntp function.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
net: dhcpv4: set sntp ip via option

Adds the option to set the ntp address ip via
dhcpv4 option 42 (Network Time Protocol Servers).

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

jukkar commented on subsys/net/lib/dhcpv4/Kconfig:
```diff
@@ -69,6 +69,12 @@ config NET_DHCPV4_VENDOR_CLASS_IDENTIFIER_STRING
 
+config NET_DHCPV4_OPTION_NTP_SERVER
+	bool "Use NTP server from DHCPv4 option and save it in the net_if"
+	help
```
We could add here
```
default y
depends on SNTP
```
so that it will set by default if SNTP support is enabled.

jukkar commented on subsys/net/lib/config/Kconfig:
```diff
@@ -234,6 +234,13 @@ config NET_CONFIG_SNTP_INIT_SERVER
 
+config NET_CONFIG_SNTP_INIT_SERVER_USE_DHCPV4_OPTION
+	bool "SNTP server to use for system clock init is set using DHCPv4 option"
+	depends on NET_DHCPV4_OPTION_NTP_SERVER
```
I think it makes sense to enable this by default if the user has enabled the NTP server option so I suggest `default y` here.

jukkar commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -11,12 +11,30 @@ LOG_MODULE_DECLARE(net_config, CONFIG_NET_CONFIG_LOG_LEVEL);
+	}
+	LOG_INF("SNTP address not set by DHCPv4, using Kconfig defaults");
+#endif /* NET_CONFIG_SNTP_INIT_SERVER_USE_DHCPV4_OPTION */
+	return sntp_simple(CONFIG_NET_CONFIG_SNTP_INIT_SERVER, CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT,
```
Quite long line that is a bit hard to read, suggest placing the timeout config value to next line.

jukkar commented on include/zephyr/net/sntp.h:
```diff
@@ -86,13 +86,31 @@ void sntp_close(struct sntp_ctx *ctx);
+ *
+ * @return 0 if ok, <0 if error (-ETIMEDOUT if timeout).
+ */
+int sntp_simple_sock(struct sockaddr *addr, socklen_t addr_len, uint32_t timeout,
```
The `sock` feels wrong here as it gives impression about socket but as this is for address setting variant, we could call this `sntp_simple_addr()` for example.

jukkar commented on subsys/net/ip/utils.c:
```diff
@@ -932,6 +932,25 @@ bool net_ipaddr_parse(const char *str, size_t str_len, struct sockaddr *addr)
 
+int net_port_set_default(struct sockaddr *addr, uint16_t default_port)
+{
+
```
Please remove extra empty line.

jukkar commented on subsys/net/ip/utils.c:
```diff
@@ -932,6 +932,25 @@ bool net_ipaddr_parse(const char *str, size_t str_len, struct sockaddr *addr)
+		LOG_ERR("Unknown address family");
+		return -EINVAL;
+	}
+	return 0;
```
Please add empty line after closing block, we try to follow that style in zephyr.

rlubos commented on subsys/net/lib/sntp/sntp_simple.c:
```diff
@@ -60,7 +45,43 @@ int sntp_simple(const char *server, uint32_t timeout, struct sntp_time *time)
+	hints.ai_protocol = 0;
+	/* 123 is the standard SNTP port per RFC4330 */
+	res = net_getaddrinfo_addr_str(server, "123", &hints, &addr);
+
```
We usually skip the empty line between function call and result check.

rlubos commented on include/zephyr/net/sntp.h:
```diff
@@ -86,13 +86,31 @@ void sntp_close(struct sntp_ctx *ctx);
+
+/**
+ * @brief Convenience function to query SNTP in one-shot fashion
+ * using a pre-initialized socket
```
Did you really mean socket here, or rather pre-initialized address struct?

rlubos commented on include/zephyr/net/sntp.h:
```diff
@@ -86,13 +86,31 @@ void sntp_close(struct sntp_ctx *ctx);
+ * @return 0 if ok, <0 if error (-ETIMEDOUT if timeout).
+ */
+int sntp_simple_addr(struct sockaddr *addr, socklen_t addr_len, uint32_t timeout,
+		       struct sntp_time *ts);
```
Please fix alignment.

rlubos commented on subsys/net/lib/sntp/sntp_simple.c:
```diff
@@ -8,34 +8,19 @@
 
-int sntp_simple(const char *server, uint32_t timeout, struct sntp_time *time)
+static int sntp_simple_helper(struct sockaddr *addr, socklen_t addr_len, uint32_t timeout,
+		       struct sntp_time *ts)
```
Please fix alignment.

rlubos commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -11,12 +11,30 @@ LOG_MODULE_DECLARE(net_config, CONFIG_NET_CONFIG_LOG_LEVEL);
+#ifdef CONFIG_NET_CONFIG_SNTP_INIT_SERVER_USE_DHCPV4_OPTION
+	struct net_if *iface = net_if_get_default();
+
+	if (iface->config.dhcpv4.ntp_addr.s_addr != 0) {
```
Could use a dedicated `net_ipv4_is_addr_unspecified()` here.

rlubos commented on subsys/net/lib/config/init_clock_sntp.c:
```diff
@@ -11,12 +11,30 @@ LOG_MODULE_DECLARE(net_config, CONFIG_NET_CONFIG_LOG_LEVEL);
+	LOG_INF("SNTP address not set by DHCPv4, using Kconfig defaults");
+#endif /* NET_CONFIG_SNTP_INIT_SERVER_USE_DHCPV4_OPTION */
+	return sntp_simple(CONFIG_NET_CONFIG_SNTP_INIT_SERVER,
+				CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT, tm);
```
Please fix alignment

rlubos reviewed (commented):
Some nitpicking, mostly about the formating.
